src/Lights/Shadows/shadowGenerator.ts

```diff
@@ -726,10 +726,16 @@ export class ShadowGenerator implements IShadowGenerator {
             this._shadowMap.renderList = [];
         }
 
-        this._shadowMap.renderList.push(mesh);
-
-        if (includeDescendants) {
-            this._shadowMap.renderList.push(...mesh.getChildMeshes());
+        var renderListMap: Record<string, undefined> = {};
+        this._shadowMap.renderList.forEach(
+            (renderedMesh) => { renderListMap[renderedMesh.name] = undefined; },
+        );
+
+        var meshes = includeDescendants ? [mesh, ...mesh.getChildMeshes()] : [mesh];

```

Comment:
In order to avoid ... (which will again generate a full traversal and a lot of GC) we can simply do something
```
this._shadowMap.renderList.push(mesh);
if (!includeDescendants) {
   return;
}

for (var addedMesh of mesh.getChildMeshes()) {
            if (!renderListMap.hasOwnProperty(addedMesh.name)) {
                this._shadowMap.renderList.push(addedMesh);
            }
        }
```

Reply:
removed spreading of children array and moved check for initial mesh into separate if statement